File: site/data-types/lists/insert-remove-size-get-single.js
'use strict';

function my_insert(index,element) {
    event_counter++;
    list.splice(index, 0, element);
}

function my_remove(index) {
	event_counter++;
	var del = list.splice(index,1);
	return del[0];
}

function my_size() {
    event_counter++;
    var s=list.length;
    return s;
}

function my_get(index) {
	event_counter++;
	var s = list[index];
	return s;
}

var trace_size = process.argv[2] || 100;
var max_size = process.argv[3] || 20;
var max_elem = 1000;
var list= [];
var event_counter=1;
var elem;
var bit;
var sz;

while(event_counter<=trace_size){
	bit = 0;
    for(elem = 1; elem <= max_size; elem++) {
		sz = my_size();
		if (bit==0){
			my_insert(0,elem);
			bit++;
		} else if (bit==1){
			my_insert(sz,elem);
			bit++;
		} else {
			my_insert(Math.floor(sz/2),elem);
			bit=0;
		}
		for (i = 0; i <= sz; i++){
			my_get(i);
		}
    }
	bit=0;
    for (elem = 1; elem <= max_size; elem++) {
		sz = my_size();
		if (bit==0){
			my_remove(0);
			bit++;
		} else if (bit==1){
			my_remove(sz-1);
			bit++;
		} else {
			my_remove(Math.floor(sz/2));
			bit=0;
		}
		for (i = 0; i < sz-1; i++){
			my_get(i);
		}
    }
}

File: site/data-types/lists/insert-remove-size-get-single.pl
:- module(spec, [(trace_expression/2), (match/2)]).
:- use_module(monitor(deep_subdict)).
:- use_module(library(clpr)).
match(_event, insert_et(Index, Elem)) :- deep_subdict(_event, _{event:"func_pre", name:"my_insert", args:[Index, Elem]}).
match(_event, remove_et(Index, Elem)) :- deep_subdict(_event, _{event:"func_post", name:"my_remove", args:[Index], res:Elem}).
match(_event, size_et(Size)) :- deep_subdict(_event, _{event:"func_post", name:"my_size", args:[], res:Size}).
match(_event, get_et(Index, Elem)) :- deep_subdict(_event, _{event:"func_post", name:"my_get", args:[Index], res:Elem}).
match(_event, relevant_et) :- match(_event, insert_et(_, _)).
match(_event, relevant_et) :- match(_event, remove_et(_, _)).
match(_event, relevant_et) :- match(_event, size_et(_)).
match(_event, relevant_et) :- match(_event, get_et(_, _)).
match(_event, add_rm_get_et) :- match(_event, insert_et(_, _)).
match(_event, add_rm_get_et) :- match(_event, remove_et(_, _)).
match(_event, add_rm_get_et) :- match(_event, get_et(_, _)).
match(_event, insert_in_bounds_et(Size)) :- match(_event, insert_et(Index, _)),
	{((Index >= 0),(Index =< Size))}.
match(_event, remove_in_bounds_et(Size)) :- match(_event, remove_et(Index, _)),
	{((Index >= 0),(Index < Size))}.
match(_event, get_in_bounds_et(Size)) :- match(_event, get_et(Index, _)),
	{((Index >= 0),(Index < Size))}.
match(_event, get_size_et(Size)) :- match(_event, size_et(Size)).
match(_event, get_size_et(Size)) :- match(_event, get_in_bounds_et(Size)).
match(_event, not_insert_et) :- not(match(_event, insert_et(_, _))).
match(_event, increased_et(I)) :- match(_event, insert_et(Index, _)),
	{(Index =< I)}.
match(_event, decreased_et(I)) :- match(_event, remove_et(Index, _)),
	{(Index < I)}.
match(_event, irrelevant_modification_et(I)) :- match(_event, insert_et(Index, _)),
	{(Index > I)}.
match(_event, irrelevant_modification_et(I)) :- match(_event, remove_et(Index, _)),
	{(Index > I)}.
match(_event, irrelevant_get_et(I)) :- match(_event, get_et(Index, _)),
	{(Index =\= I)}.
match(_event, irrelevant_et(I)) :- match(_event, irrelevant_modification_et(I)).
match(_event, irrelevant_et(I)) :- match(_event, irrelevant_get_et(I)).
match(_event, any_et) :- deep_subdict(_event, _{}).
match(_event, none_et) :- not(match(_event, any_et)).
trace_expression('Main', Main) :- (Main=((relevant_et>>clos((app(CheckIndex, [0])/\((add_rm_get_et>>CheckElem);1))));1)),
	(CheckIndex=gen([size], (star(get_size_et(var(size)))*((insert_in_bounds_et(var(size))*app(CheckIndex, [(var(size) + 1)]))\/(remove_in_bounds_et(var(size))*app(CheckIndex, [(var(size) - 1)])))))),
	(CheckElem=(star(not_insert_et)*var(index, var(elem, (insert_et(var(index), var(elem))*(app(GetElem, [var(index), var(elem)])/\CheckElem)))))),
	(GetElem=gen([index, elem], (((((irrelevant_et(var(index))\/get_et(var(index), var(elem)))*app(GetElem, [var(index), var(elem)]))\/(increased_et(var(index))*app(GetElem, [(var(index) + 1), var(elem)])))\/(decreased_et(var(index))*app(GetElem, [(var(index) - 1), var(elem)])))\/(remove_et(var(index), var(elem))*1)))).
